exporter/datadogexporter/factory.go

```diff
@@ -99,23 +100,24 @@ func createMetricsExporter(
 		return nil, err
 	}
 
+	ctx, cancel := context.WithCancel(ctx)
 	var pushMetricsFn exporterhelper.PushMetrics
 
 	if cfg.OnlyMetadata {
-		pushMetricsFn = func(context.Context, pdata.Metrics) (int, error) {
-			// if only sending metadata ignore all metrics
+		pushMetricsFn = func(_ context.Context, md pdata.Metrics) (int, error) {
+			// only sending metadata use only metrics
+			once := cfg.OnceMetadata()
+			once.Do(func() {
+				attrs := pdata.NewAttributeMap()
+				if md.ResourceMetrics().Len() > 0 {
```

Comment:
There may be one weird edge case here: what if, during the invocation of `pushMetricsFn`, there may be sometimes (but not always) no `ResourceMetrics` in `md`?. This could lead to some weird behavior where upon restarting the collector, metadata may change depending on whether  the first`md` object passed to the exporter contains reosurce metrics or not, which may be confusing.

I don't know if there's a good solution to this though. You could wait until you find a `md` with resource metrics, but that means metadata isn't sent as fast as possible, which would add its own set of problems.

Reply:
I think this can't happen if the pipeline configuration doesn't change. The processors are initialized before the receivers and they gather metadata at startup, if they fail to do so the Collector process stops. So, (if my understanding of the code is correct) it can either be that the Collector fails to start or that the metadata is gathered correctly and applied to every payload.